collaboration/extending-kyma-cli.md

```diff
@@ -0,0 +1,351 @@
+
+## Architecture
+
+We should follow the modular approach with each component having its own library. This library can be used inside the kyma cli for acessing various components. The library contains all the implementation. The kyma cli should invoke this library. We can take inspiration from hydroform and implement it in similar way. Currently we have following modules in kyma:
+
+  1. Functions
+  2. Service Catalog
+  3. Events
+  4. Application Connector
+  5. APIs
+
+## Command structure
+
+The command structure used by cloud providers is following:
+
+```bash
+<cliName> <functionality> <action>
+#example aws s3 ls
+
+# Functions
+kyma lambda <action>
+
+# Events
+kyma events <action>
+
+# Application Connector
+kyma application <events>
+
+# Apis
+kyma apis <events>
+```
+
+### Functions
+
+For the serverless component we should start with supporting following commands
+
+#### Local workspace setup
+
+For the local workspace setup developer should be provied with a project containing basic working function example. This should be deployable on to an existing kyma cluster using the kyma cli. It should also be possible to test this code locally by means of make files. User should be provded with some readme where he can understand the project structure and also the configuration examples. It should also contain commands to deploy and debug the function on to kyma cluster and also how to run locally.
+
+##### Creation of project locally
+
+This could be used for the user who is new to kyma and kubernetes in general. He should be able to have an example function created easily. We can have following commands.
+
+```bash
+  kyma function init <function-name> --language <language-name>
+
+  kyma function init <function-name> --language <language-name> -n <namespace> -p <path>
+
+  kyma function init <function-name> --language <language-name> --from-git <path-to-function-code>
+```
+
+> The namespace should be optional. When not passed it can be set to `default`. When the user has path to the function code (residing in git) we should generate the scaffolding around it (if its not existing already).
+When path `-p` is passed then it should create the project in the designated path, otherwise, it uses the current directory.
+The language corresponds the language in which the function code would be written.
+
+This should create directoy structure with following content based on the language passed (assuming vscode as the IDE).:
+
+```bash
+  <function-name>
+  ├── .vscode
+  │   └── launch.json
+  ├── deployment
+  │   └── deployment.yaml
+  ├── local
+  │   └── <server-code.extension>
+  └── Readme.md
+  └── code
+    ├── config.yaml
+    ├── <function-code.extension>
+    └── <dependecies.extension>
+```
+
+* launch.json consists of settings for the local debugger on vscode
+* deployment.yaml consists of the yaml of the function cr that has been deployed on the kyma cluster. This would be generated once function has been deployed.
```

Comment:
Isn't deployment.yaml still misleading? Maybe function.yaml?

Reply:
The file will not be used later, it is just for visibility or a dry-run?

Reply:
It is to show the user what has been deployed onto the cluster.